index.js

```diff
@@ -200,9 +275,250 @@ function fastifyMultipart (fastify, options, done) {
-    return this.raw[kMultipart] || false
+  function handleMultipart (opts = {}) {
+    if (!this.isMultipart()) {
+      throw new InvalidMultipartContentTypeError()
+    }
+
+    let worker
+    let lastValue
+
+    // only one file / field can be processed at a time
+    // "null" will close the consumer side
+    const ch = (val) => {
+      if (typeof val === 'function') {
+        worker = val
+      } else {
+        lastValue = val
+      }
+      if (worker && lastValue !== undefined) {
+        worker(lastValue)
+        worker = undefined
+        lastValue = undefined
+      }
+    }
+    const parts = () => {
+      return new Promise((resolve, reject) => {
+        ch((val) => {
+          if (val instanceof Error) return reject(val)
+          resolve(val)
+        })
+      })
+    }
+
+    const body = {}
+    let lastError = null
+    const request = this.raw
+    const busboyOptions = deepmerge.all([
+      { headers: request.headers },
+      options,
+      opts
+    ])
+
+    const bb = busboy(busboyOptions)
+
+    request.on('close', cleanup)
+
+    bb
+      .on('field', onField)
+      .on('file', onFile)
+      .on('close', cleanup)
+      .on('error', onEnd)
+      .on('finish', onEnd)
+
+    bb.on('partsLimit', function () {
+      onError(new PartsLimitError())
+    })
+
+    bb.on('filesLimit', function () {
+      onError(new FilesLimitError())
+    })
+
+    bb.on('fieldsLimit', function () {
+      onError(new FieldsLimitError())
+    })
+
+    request.pipe(bb)
+
+    function onField (name, fieldValue, fieldnameTruncated, valueTruncated) {
+      // don't overwrite prototypes
+      if (getDescriptor(Object.prototype, name)) {
+        onError(new PrototypeViolationError())
+        return
+      }
+
+      const value = {
+        fieldname: name,
+        value: fieldValue,
+        fieldnameTruncated,
+        valueTruncated,
+        fields: body
+      }
+
+      if (body[name] === undefined) {
+        body[name] = value
+      } else if (Array.isArray(body[name])) {
+        body[name].push(value)
+      } else {
+        body[name] = [body[name], value]
+      }
+
+      ch(value)
+    }
+
+    function onFile (name, file, filename, encoding, mimetype) {
+      // don't overwrite prototypes
+      if (getDescriptor(Object.prototype, name)) {
+        // ensure that stream is consumed, any error is suppressed
+        sendToWormhole(file)
+        onError(new PrototypeViolationError())
+        return
+      }
+
+      const value = {
+        fieldname: name,
+        filename,
+        encoding,
+        mimetype,
+        file,
+        fields: body,
+        _buf: null,
+        buffer () {
+          if (this._buf) {
+            return Promise.resolve(this._buf)
+          }
+          return new Promise((resolve, reject) => {
+            pump(this.file, concat((buf) => {
+              this._buf = buf
+              resolve(this._buf)
+            })).catch(reject)
+          })
```

Comment:
I would avoid breaking the promise chain for this. Maybe we can use `this.file` async iterator instead?

Reply:
Good point.